File: status.py
"""
Pipeline status dashboard — read-only monitor for orchestration protocol.
stdlib only. Run: python status.py [poll_interval_seconds]
"""

import os
import sqlite3
import sys
import time


def status(db_path=".claude/orchestrator/orchestrator.db", interval=3):
    conn = _open_ro(db_path)

    # Track last audit_log id to detect writes
    last_audit_id = _max_audit_id(conn)
    conn.close()

    while True:
        conn = _open_ro(db_path)
        pipelines = _fetch_all(conn)
        current_audit_id = _max_audit_id(conn)
        conn.close()

        _render(pipelines, current_audit_id, last_audit_id)

        if current_audit_id is not None and current_audit_id != last_audit_id:
            # Write detected — cooldown then immediate refresh
            time.sleep(2)
            last_audit_id = current_audit_id
        else:
            last_audit_id = current_audit_id
            time.sleep(interval)


def _open_ro(db_path):
    """Open read-only connection. Returns None if DB doesn't exist yet."""
    if not os.path.exists(db_path):
        return None
    uri = "file:{}?mode=ro".format(db_path.replace("\\", "/"))
    try:
        conn = sqlite3.connect(uri, uri=True)
        conn.row_factory = sqlite3.Row
        return conn
    except sqlite3.OperationalError:
        return None


def _max_audit_id(conn):
    if conn is None:
        return None
    try:
        row = conn.execute("SELECT MAX(id) FROM audit_log").fetchone()
        return row[0]
    except sqlite3.OperationalError:
        return None


def _fetch_all(conn):
    if conn is None:
        return []
    try:
        rows = conn.execute("""
            SELECT request_id, agent, stage, revision,
                   reason, approval_status, rejection_reason,
                   updated_at
            FROM pipeline_state
            ORDER BY updated_at DESC
        """).fetchall()
        return [dict(r) for r in rows]
    except sqlite3.OperationalError:
        return []


# ── Display ────────────────────────────────────────────────────

STAGE_ICONS = {
    "request_submitted":       "○",  # ○
    "conflict_analysis_done":  "◐",  # ◐
    "boundary_analysis_done":  "◑",  # ◑
    "logic_analysis_done":     "◕",  # ◕
    "approved":                "✔",  # ✔
    "rejected":                "✘",  # ✘
    "modifying":               "⚒",  # ⚒
    "self_review_done":        "◉",  # ◉
    "completion_submitted":    "⏳",  # ⏳
    "completed":               "★",  # ★
    "lock_released":           "☑",  # ☑
}

STAGE_WIDTH = 22


def _render(pipelines, audit_id, last_audit_id):
    _clear()
    changed = audit_id is not None and last_audit_id is not None and audit_id != last_audit_id

    print(" Pipeline Status Dashboard")
    print("=" * 80)
    print(" DB: .claude/orchestrator/orchestrator.db | "
          "audit #{:>6} | {}".format(
              audit_id or 0,
              "[CHANGED — refreshed after 2s cooldown]" if changed
              else "polling every few seconds...",
          ))
    print("=" * 80)

    if not pipelines:
        print("\n  (no pipelines yet — waiting for worker to init)\n")
        return

    for p in pipelines:
        icon = STAGE_ICONS.get(p["stage"], "?")
        stage_label = "{}{}".format(icon, p["stage"])
        reason = (p["reason"] or "")[:48]

        # Build status tags
        tags = ""
        if p["stage"] == "rejected":
            tags = " REASON: {}".format((p["rejection_reason"] or "")[:40])
        elif p["stage"] == "approved":
            tags = " APPROVED"
        elif p["stage"] == "completed":
            tags = " AWAITING LOCK RELEASE"

        print(" {:<6} | {:<24} | rev {:>2} | {:<50} | {}".format(
            p["agent"][:6],
            stage_label[:24],
            p["revision"],
            reason[:50],
            p["updated_at"] or "",
        ))
        if tags:
            print("        | {:<24} |       | {}".format("", tags))

    print("\n" + "-" * 80)
    stage_counts = {}
    for p in pipelines:
        stage_counts[p["stage"]] = stage_counts.get(p["stage"], 0) + 1
    summary = " | ".join(
        "{}:{}{}".format(STAGE_ICONS.get(s, "?"), c, s)
        for s, c in sorted(stage_counts.items())
    )
    print(" {}".format(summary))
    print(" Ctrl+C to exit\n")


def once(db_path=".claude/orchestrator/orchestrator.db"):
    """Print status once and exit — for agent invocation."""
    conn = _open_ro(db_path)
    pipelines = _fetch_all(conn)
    audit_id = _max_audit_id(conn)
    conn.close()
    _render(pipelines, audit_id, None)
    return pipelines


def _clear():
    os.system("cls" if os.name == "nt" else "clear")


if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "--once":
        db = sys.argv[2] if len(sys.argv) > 2 else ".claude/orchestrator/orchestrator.db"
        once(db)
    else:
        db = sys.argv[1] if len(sys.argv) > 1 else ".claude/orchestrator/orchestrator.db"
        sec = int(sys.argv[2]) if len(sys.argv) > 2 else 3
        try:
            status(db, sec)
        except KeyboardInterrupt:
            print()

File: src/bin/status.rs
//! Pipeline status dashboard — read-only monitor.
//!
//! Replaces Python `status.py`.  Reads pipeline_state and audit_log,
//! renders a terminal dashboard.  Supports `--once` for agent invocation
//! and polling mode with configurable interval.

use clap::Parser;
use rusqlite::Connection;
use std::collections::HashMap;
use std::thread;
use std::time::Duration;

#[derive(Parser)]
#[command(name = "status", about = "Pipeline status dashboard")]
struct Args {
    #[arg(default_value = ".claude/orchestrator/orchestrator.db")]
    db_path: String,

    #[arg(long)]
    once: bool,

    #[arg(default_value = "3")]
    interval: u64,
}

fn main() {
    let args = Args::from_args();
    let mut last_audit_id: i64 = 0;

    loop {
        let conn = open_ro(&args.db_path);
        let (pipelines, current_audit_id) = if let Some(ref c) = conn {
            (fetch_all(c), max_audit_id(c))
        } else {
            (vec![], None)
        };

        let changed = current_audit_id.map_or(false, |id| id != last_audit_id);
        render(&pipelines, current_audit_id, changed);
        last_audit_id = current_audit_id.unwrap_or(0);

        if args.once {
            break;
        }
        if changed {
            thread::sleep(Duration::from_secs(2));
        } else {
            thread::sleep(Duration::from_secs(args.interval));
        }
    }
}

fn open_ro(db_path: &str) -> Option<Connection> {
    if !std::path::Path::new(db_path).exists() {
        return None;
    }
    let uri = format!("file:{}?mode=ro", db_path.replace('\\', "/"));
    Connection::open_with_flags(&uri, rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY).ok()
}

fn max_audit_id(conn: &Connection) -> Option<i64> {
    conn.query_row("SELECT COALESCE(MAX(id), 0) FROM audit_log", [], |row| {
        row.get(0)
    })
    .ok()
}

fn fetch_all(conn: &Connection) -> Vec<HashMap<String, String>> {
    let mut stmt = match conn.prepare(
        "SELECT request_id, agent, stage, revision, reason_json, approval_status, rejection_reason, updated_at, human_intervention
         FROM pipeline_state ORDER BY updated_at DESC",
    ) {
        Ok(s) => s,
        Err(_) => return vec![],
    };
    let rows = stmt
        .query_map([], |row| {
            let mut map = HashMap::new();
            for (i, col) in [
                "request_id", "agent", "stage", "revision", "reason_json",
                "approval_status", "rejection_reason", "updated_at", "human_intervention",
            ]
            .iter()
            .enumerate()
            {
                let val: Option<String> = row.get(i).ok().flatten();
                if let Some(v) = val {
                    map.insert(col.to_string(), v);
                }
            }
            Ok(map)
        })
        .ok();
    match rows {
        Some(r) => r.filter_map(|r| r.ok()).collect(),
        None => vec![],
    }
}

fn render(pipelines: &[HashMap<String, String>], audit_id: Option<i64>, changed: bool) {
    print!("\x1B[2J\x1B[H"); // clear screen

    let id = audit_id.unwrap_or(0);
    let changed_str = if changed {
        " [CHANGED — refreshed after 2s cooldown]"
    } else {
        "polling..."
    };

    println!(" Pipeline Status Dashboard");
    println!("{:=<80}", "");
    println!(
        " DB: .claude/orchestrator/orchestrator.db | audit #{:>6} | {}",
        id, changed_str
    );
    println!("{:=<80}", "");

    if pipelines.is_empty() {
        println!("\n  (no pipelines yet — waiting for worker to init)\n");
        return;
    }

    let icons: HashMap<&str, &str> = HashMap::from([
        ("init", "○"),
        ("orchestrator_gate", "◐"),
        ("worker_modify", "⚒"),
        ("reviewer_check", "◉"),
        ("orchestrator_arbiter", "◕"),
        ("verified", "✔"),
        ("rejected", "✘"),
        ("completed", "★"),
        ("lock_released", "☑"),
    ]);

    for p in pipelines {
        let stage = p.get("stage").map(|s| s.as_str()).unwrap_or("?");
        let icon = icons.get(stage).unwrap_or(&"?");
        let agent = p.get("agent").map(|a| &a[..a.len().min(6)]).unwrap_or("");
        let rev = p.get("revision").map(|r| r.as_str()).unwrap_or("?");
        let reason = p
            .get("reason_json")
            .map(|r| &r[..r.len().min(48)])
            .unwrap_or("");

        let hi_flag = if p
            .get("human_intervention")
            .map_or(false, |h| h.contains("needs_human"))
        {
            " [NEEDS HUMAN]"
        } else {
            ""
        };

        println!(
            " {:<6} | {}{:<20} | rev {:>2} | {:<48} | {}{}",
            agent,
            icon,
            stage,
            rev,
            reason,
            p.get("updated_at").map(|u| u.as_str()).unwrap_or(""),
            hi_flag,
        );

        if stage == "rejected" {
            if let Some(r) = p.get("rejection_reason") {
                println!(
                    "        | {:<22} |       | REJECTED: {}",
                    "",
                    &r[..r.len().min(40)]
                );
            }
        }
    }
    println!("{:-<80}", "");
}
